### PR TITLE
added UserContext, and login/logout features

### PR DESCRIPTION
**Pull request by tmurphy3**
## Related Issues (include the `#`)
#68 

## Description of Changes Made
wrapped app with state provider

### CODE REVIEW

File: front-end/src/Pages/Profile.js [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+import React, { useContext } from "react";
 import ProjectList from "../Components/ProjectList";
+import { UserContext } from "../UserContext";
 
 function Profile() {
+  const { user } = useContext(UserContext);
+
   //the projects will be passed props of what users profile we are on
+
   return (
```
Comment (midastouchprd) [inline front-end/src/Pages/Profile.js]: you need to define a state for null user.  should redirect to login

**Review by midastouchprd — CHANGES_REQUESTED** [review]

File: front-end/src/Pages/Profile.js (hunk 1 above)
Comment (tmurphy3) [reply front-end/src/Pages/Profile.js]: profile will never render unless a user is defined. I have the navigation bar conditionally rendering login OR logout/profile... also when login or logout is submitted I route the app back to the home page

File: front-end/src/App.js [hunk 2]
```diff
@@ -1,29 +1,56 @@
...
+              )}
+              <Route
+                exact
+                path="/login"
+                render={({ match }) => <Login match={match} />}
+              />
+              <Route
+                exact
+                path="/profile/:id"
+                render={({ match }) => <Profile match={match} />}
```
Comment (midastouchprd) [inline front-end/src/App.js]: Make a `<SecureRoute />` component for these that'll automatically handle the authentication redirection

https://ui.dev/react-router-v4-protected-routes-authentication/

**Review by midastouchprd — CHANGES_REQUESTED** [review]

File: front-end/src/App.js [hunk 3]
```diff
@@ -1,29 +1,56 @@
...
+import Login from "./Pages/Login";
+import Profile from "./Pages/Profile";
 
 export default function App() {
+  const [user, setUser] = useState(null);
+
+  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
+
   return (
     <HashRouter basename="/">
```
Comment (midastouchprd) [inline front-end/src/App.js]: I'd use BrowserRouter instead of HashRouter

**Review by midastouchprd — CHANGES_REQUESTED** [review]

File: front-end/src/App.js (hunk 3 above)
Comment (midastouchprd) [reply front-end/src/App.js]: one hash router won't keep your browser history correct so you'll get things like

profile/1/#/about/profile

Comment (midastouchprd) [reply front-end/src/App.js]: it gives the illusion of url protection but it really is just forcing you to do extra logic for it.

Comment (midastouchprd) [reply front-end/src/App.js]: you don't want a user to be on `/account/1` and be shown a login page.

you want them redirected to the login page.

File: front-end/src/App.js [hunk 4]
```diff
@@ -1,29 +1,56 @@
...
-      <div>
+      <UserContext.Provider value={providerValue}>
```
Comment (midastouchprd) [inline front-end/src/App.js]: GREAT JOB USING CONTEXT THOUGH!

File: front-end/src/App.js (hunk 3 above)
Comment (tmurphy3) [reply front-end/src/App.js]: Okay, I am using browser router now. if the user is logged in they will be shown their profile if they are on the route /myprofile/number. but if they are not logged in and they somehow get to that route, the login page will be rendered instead.

File: front-end/src/Components/SecureRoutes.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import React, { useContext } from "react";
```
Comment (midastouchprd) [inline front-end/src/Components/SecureRoutes.js:1]: not exactly what i meant but clever.

**Review by midastouchprd — APPROVED** [review]

**Pull request merged**